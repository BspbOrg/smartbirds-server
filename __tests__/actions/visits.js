/* eslint-env jest */
/* global setup */

const userFactory = require('../../__utils__/factories/userFactory')
const visitFactory = require('../../__utils__/factories/visitFactory')

const ROLES_PERMISSIONS = [
  ['guest', null, {
    list: false,
    view: false,
    edit: false,
    create: false,
    delete: false
  }],
  ['user', { role: 'user' }, {
    list: true,
    view: true,
    edit: false,
    create: false,
    delete: false
  }],
  ['birds moderator', { role: 'moderator', forms: { formBirds: true } }, {
    list: true,
    view: true,
    edit: false,
    create: false,
    delete: false
  }],
  ['cbm moderator', { role: 'moderator', forms: { formCBM: true } }, {
    list: true,
    view: true,
    edit: true,
    create: true,
    delete: true
  }],
  ['birds and cbm moderator', { role: 'moderator', forms: { formBirds: true, formCBM: true } }, {
    list: true,
    view: true,
    edit: true,
    create: true,
    delete: true
  }],
  ['org admin', { role: 'org-admin' }, {
    list: true,
    view: true,
    edit: true,
    create: true,
    delete: true
  }],
  ['admin', { role: 'admin' }, {
    list: true,
    view: true,
    edit: true,
    create: true,
    delete: true
  }]
]

const generateApiPeriod = (year) => ({
  early: {
    start: `${year}-01-01`,
    end: `${year}-01-02`
  },
  late: {
    start: `${year}-02-01`,
    end: `${year}-02-02`
  }
})

const testPermissions = (action, permission, allowedTest, forbiddenTest) => {
  const allowed = ROLES_PERMISSIONS.filter(([_, __, permissions]) => permissions[permission])
  const forbidden = ROLES_PERMISSIONS.filter(item => !allowed.includes(item))
  const runWrapper = (test) => async (_, userParams) => {
    const user = userParams != null ? await userFactory(setup.api, userParams) : null

    await test((params) => user
      ? setup.runActionAs(action, params, user.email)
      : setup.runAction(action, params)
    )
  }

  if (allowed.length || forbidden.length) {
    describe(permission, () => {
      if (allowed.length) {
        test.each(allowed)(`%s can ${permission}`, runWrapper(allowedTest))
      }

      if (forbidden.length) {
        test.each(forbidden)(`%s cannot ${permission}`, runWrapper(forbiddenTest))
      }
    })
  }
}

describe('Action: visits', () => {
  const missingYear = 1982
  let visit

  beforeEach(async () => {
    visit = await visitFactory(setup.api)

    await setup.api.models.visit.destroy({ where: { year: missingYear } })
  })

  testPermissions('visit:list', 'list', async (runAction) => {
    const response = await runAction()

    expect(response).not.toEqual(expect.objectContaining({ error: expect.anything() }))
    expect(response).toEqual(expect.objectContaining({
      data: expect.arrayContaining([
        expect.objectContaining({ year: visit.year })
      ])
    }))
  }, async (runAction) => {
    const response = await runAction()

    expect(response.error).toMatchSnapshot()
    expect(response.data).toBeFalsy()
  })

  testPermissions('visit:view', 'view', async (runAction) => {
    const response = await runAction({ year: visit.year })

    expect(response).not.toEqual(expect.objectContaining({ error: expect.anything() }))
    expect(response).toEqual(expect.objectContaining({
      data: expect.objectContaining({ year: visit.year })
    }))
  }, async (runAction) => {
    const response = await runAction({ year: visit.year })

    expect(response.error).toMatchSnapshot()
    expect(response.data).toBeFalsy()
  })

  testPermissions('visit:edit', 'edit', async (runAction) => {
    const periods = generateApiPeriod(visit.year)

    const response = await runAction({
      year: visit.year,
      ...periods
    })

    expect(response).not.toEqual(expect.objectContaining({ error: expect.anything() }))
    expect(response).toEqual(expect.objectContaining({
      data: expect.objectContaining({
        year: visit.year,
        ...periods
      })
    }))
  }, async (runAction) => {
    const response = await runAction({
      year: visit.year,
      ...generateApiPeriod(visit.year)
    })

    expect(response.error).toMatchSnapshot()
    expect(response.data).toBeFalsy()
  })

  testPermissions('visit:edit', 'create', async (runAction) => {
    const periods = generateApiPeriod(missingYear)

    const response = await runAction({
      year: missingYear,
      ...periods
    })

    expect(response).not.toEqual(expect.objectContaining({ error: expect.anything() }))
    expect(response).toEqual(expect.objectContaining({
      data: expect.objectContaining({
        year: missingYear,
        ...periods
      })
    }))
  }, async (runAction) => {
    const response = await runAction({
      year: missingYear,
      ...generateApiPeriod(missingYear)
    })

    expect(response.error).toMatchSnapshot()
    expect(response.data).toBeFalsy()
  })

  testPermissions('visit:delete', 'delete', async (runAction) => {
    const response = await runAction({ year: visit.year })

    expect(response).not.toEqual(expect.objectContaining({ error: expect.anything() }))
    expect(await setup.api.models.visit.findOne({ where: { year: visit.year } })).toBe(null)
  }, async (runAction) => {
    const response = await runAction({ year: visit.year })

    expect(response.error).toMatchSnapshot()
    expect(response.data).toBeFalsy()
    expect(await setup.api.models.visit.findOne({ where: { year: visit.year } })).not.toBe(null)
  })
})
