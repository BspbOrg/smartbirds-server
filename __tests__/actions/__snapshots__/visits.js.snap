// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action: visits create birds moderator cannot create 1`] = `"Error: Admin required"`;

exports[`Action: visits create guest cannot create 1`] = `"Error: Please log in to continue"`;

exports[`Action: visits create user cannot create 1`] = `"Error: Admin required"`;

exports[`Action: visits delete birds moderator cannot delete 1`] = `"Error: Admin required"`;

exports[`Action: visits delete guest cannot delete 1`] = `"Error: Please log in to continue"`;

exports[`Action: visits delete user cannot delete 1`] = `"Error: Admin required"`;

exports[`Action: visits edit birds moderator cannot edit 1`] = `"Error: Admin required"`;

exports[`Action: visits edit guest cannot edit 1`] = `"Error: Please log in to continue"`;

exports[`Action: visits edit user cannot edit 1`] = `"Error: Admin required"`;

exports[`Action: visits list guest cannot list 1`] = `"Error: Please log in to continue"`;

exports[`Action: visits view guest cannot view 1`] = `"Error: Please log in to continue"`;
